lib/flapjack/gateways/pager_duty.rb

```diff
@@ -0,0 +1,411 @@
+          Flapjack::Gateways::PagerDuty.send_pagerduty_event(
+            :service_key  => address,
+            :incident_key => check.name,
+            :event_type   => pagerduty_type,
+            :description  => msg,
+            :details      => {'HOSTNAME' => host_name,
+                               'SERVICE' => service_name}
+          )
+
+          Flapjack.logger.info "Sent alert successfully: #{alert.to_s}"
+        rescue => e
+          Flapjack.logger.error "Error generating or dispatching PagerDuty message: #{e.class}: #{e.message}\n" +
+            e.backtrace.join("\n")
+          Flapjack.logger.debug "Alert that could not be processed: \n" + alert.inspect
+        end
+
+      end
+
+      class AckFinder
+
+        SEM_PAGERDUTY_ACKS_RUNNING = 'sem_pagerduty_acks_running'
+
+        def initialize(opts = {})
+          @lock = opts[:lock]
+          @config = opts[:config]
+
+          @initial = true
+
+          Flapjack.logger.debug("New PagerDuty::AckFinder pikelet with the following options: #{@config.inspect}")
+
+          # TODO: only clear this if there isn't another PagerDuty gateway instance running
+          # or better, include an instance ID in the semaphore key name
+          Flapjack.redis.del(SEM_PAGERDUTY_ACKS_RUNNING)
+        end
+
+        def start
+          until Flapjack::Gateways::PagerDuty.test_pagerduty_connection
+            Flapjack.logger.error("Can't connect to the PagerDuty API, retrying after 10 seconds")
+            Kernel.sleep(10)
+          end
+
+          loop do
+            @lock.synchronize do
+              # ensure we're the only instance of the PagerDuty acknowledgement check running (with a naive
+              # timeout of five minutes to guard against stale locks caused by crashing code) either in this
+              # process or in other processes
+              if Flapjack.redis.setnx(SEM_PAGERDUTY_ACKS_RUNNING, 'true') == 0
+                Flapjack.logger.debug("skipping looking for acks in PagerDuty as this is already happening")
+              else
+                Flapjack.redis.expire(SEM_PAGERDUTY_ACKS_RUNNING, 300)
+                find_pagerduty_acknowledgements
+                Flapjack.redis.del(SEM_PAGERDUTY_ACKS_RUNNING)
+              end
+            end
+
+            Kernel.sleep 10
+          end
+        end
+
+        def stop_type
+          :exception
+        end
+
+        def find_pagerduty_acknowledgements
+          Flapjack.logger.debug("looking for acks in PagerDuty for unack'd problems")
+
+          time = Time.now
+
+          unacked_failing_checks = []
+
+          Flapjack::Data::Check.lock(
+            Flapjack::Data::ScheduledMaintenance,
+            Flapjack::Data::UnscheduledMaintenance) do
+
+            unacked_failing_checks = Flapjack::Data::Check.
+              intersect(:failing => true).reject do |check|
+
+              check.in_scheduled_maintenance?(time) ||
+                check.in_unscheduled_maintenance?(time)
```

Comment:
We don't want to exclude failing checks that are in scheduled maintenance from this, I don't think?